rust/src/passwd.rs

```diff
@@ -182,97 +182,87 @@ pub fn concat_fs_content(
             Ok(())
         })
-        .context("failed to write /etc/passwd")?;
+        .with_context(|| format!("failed to write /etc/{}", &target))?;
 
     Ok(())
 }
 
-fn concat_group_files(rootfs: &openat::Dir, prev_root: &gio::File) -> Result<()> {
-    use gio::prelude::*;
-
-    let orig_etc_content = prev_root.resolve_relative_path("usr/etc/group");
-    let orig_usrlib_content = prev_root.resolve_relative_path("usr/lib/group");
-
-    if orig_etc_content.is_none() && orig_usrlib_content.is_none() {
-        // This could actually happen after we transition to systemd-sysusers;
-        // we won't have a need for preallocated user data in the tree.
-        return Ok(());
+fn passwd_append_unique(
+    src: &gio::File,
+    seen: &mut HashSet<String>,
+    dest: &mut BufWriter<File>,
+) -> Result<()> {
+    let src_stream = src.read(gio::NONE_CANCELLABLE)?.into_read();
+    let mut buf_rd = BufReader::new(src_stream);
+    let entries = nameservice::passwd::parse_passwd_content(&mut buf_rd)?;
+    for passwd in entries {
+        if seen.contains(&passwd.name) {
+            continue;
+        }
+        seen.insert(passwd.name.clone());
+        passwd.to_writer(dest)?;
     }
+    Ok(())
+}
 
-    let group_append_unique =
-        |src: &gio::File, seen: &mut HashSet<String>, dest: &mut BufWriter<File>| -> Result<()> {
-            let src_stream = src.read(gio::NONE_CANCELLABLE)?.into_read();
-            let mut buf_rd = BufReader::new(src_stream);
-            let entries = nameservice::group::parse_group_content(&mut buf_rd)?;
-            for group in entries {
-                if seen.contains(&group.name) {
-                    continue;
-                }
-                seen.insert(group.name.clone());
-                group.to_writer(dest)?;
-            }
-            Ok(())
-        };
-
-    rootfs
-        .write_file_with_sync("etc/group", 0o664, |dest_bufwr| -> Result<()> {
-            let mut seen_names = HashSet::new();
-            if let Some(ref src_file) = orig_etc_content {
-                group_append_unique(src_file, &mut seen_names, dest_bufwr)
-                    .context("failed to process /usr/etc/group")?;
-            }
-            if let Some(ref src_file) = orig_usrlib_content {
-                group_append_unique(src_file, &mut seen_names, dest_bufwr)
-                    .context("failed to process /usr/lib/group")?;
-            };
-            Ok(())
-        })
-        .context("failed to write /etc/group")?;
-
+fn group_append_unique(
+    src: &gio::File,
+    seen: &mut HashSet<String>,
+    dest: &mut BufWriter<File>,
+) -> Result<()> {
+    let src_stream = src.read(gio::NONE_CANCELLABLE)?.into_read();
+    let mut buf_rd = BufReader::new(src_stream);
+    let entries = nameservice::group::parse_group_content(&mut buf_rd)?;
+    for group in entries {
+        if seen.contains(&group.name) {
+            continue;
+        }
+        seen.insert(group.name.clone());
```

Comment:
I stumbled on this before...is there really no good way to use the `.entry()` API to do a lookup on `&str` and insert with `String`?

Reply:
I think the closest is the nigthly `get_or_insert()`. TBH there are a few ways to rewrite this (via `replace` or by always inserting and checking the result), but I settled on the current explicit form for readability, as this shouldn't in an hot path.